src/agent.rs

```diff
@@ -194,6 +197,29 @@ impl Agent {
             .set_max_idle_connections_per_host(max_connections);
     }
 
+    /// Configures a custom resolver to be used by this agent. By default,
+    /// address-resolution is done by std::net::ToSocketAddrs. This allows you
+    /// to override that resolution with your own alternative. Useful for
+    /// testing and special-cases like DNS-based load balancing.
+    ///
+    /// A `Fn(&str) -> io::Result<Vec<SocketAddr>>` is a valid resolver,
+    /// passing a closure is a simple way to override. Note that you might need
+    /// explicit type `&str` on the closure argument for type inference to
+    /// succeed.
+    /// ```
+    /// use std::net::ToSocketAddrs;
+    ///
+    /// let mut agent = ureq::agent();
+    /// agent.set_resolver(|addr: &str| match addr {
+    ///    "example.com" => Ok(vec![([127,0,0,1], 8096).into()]),
+    ///    addr => addr.to_socket_addrs().map(Iterator::collect),
+    /// });
+    /// ```
+    pub fn set_resolver(&mut self, resolver: impl crate::Resolver + 'static) -> &mut Self {
```

Comment:
Why the `'static` lifetime bound? Also why does this need to be a trait object as opposed to just an object that implements the trait?

Reply:
@jsha 

> Why the `'static` lifetime bound?

Since we're taking ownership of the concrete resolver impl, it's necessary in this case. If we turn it around, the types that _don't_ implement `'static` are reference types (they live for the duration of the reference). The only way to accept reference types would be to make `Agent` generic over the `Resolver` so that the entire struct would have the lifetime of the reference.

> Also why does this need to be a trait object as opposed to just an object that implements the trait?

Isn't that the same? These two are pretty much identical:

1. `fn set_resolver<R: Resolver + 'static>(&mut self, resolver: R) -> &mut self`
2. `fn set_resolver(&mut self, resolver: impl Resolver + 'static) -> &mut self`

Compilation wise they both cause a _generic function_, and must do so since the size of `impl Resolver` isn't known. Machine code explains why: the size of the stack frame must be known before jumping to the function ([monomorphization](https://rustc-dev-guide.rust-lang.org/backend/monomorph.html)).

We could have opted for a signature accepting `Arc<dyn Resolver>`, forcing the caller to box up the concrete impl before calling. That would make for more efficient compilation, but less good ergonomics.